config/v1alpha1/types_backup.go

```diff
@@ -0,0 +1,131 @@
+	Spec BackupSpec `json:"spec"`
+	// status holds observed values from the cluster. They may not be overridden.
+	// +kubebuilder:validation:Optional
+	// +optional
+	Status BackupStatus `json:"status"`
+}
+
+type BackupSpec struct {
+	// +kubebuilder:validation:Required
+	EtcdBackupSpec EtcdBackupSpec `json:"etcdBackupSpec,omitempty"`
+}
+
+type BackupStatus struct {
+}
+
+// EtcdBackupSpec provides configuration for automated etcd backups to the cluster-etcd-operator
+type EtcdBackupSpec struct {
+
+	// Schedule defines the recurring backup schedule in Cron format
+	// every 2 hours: 0 */2 * * *
+	// every day at 3am: 0 3 * * *
+	// Setting to an empty string "" means disabling scheduled backups
+	// Default: ""
+	// TODO: Define CEL validation for the cron format
+	// See if the upstream CronJob has CEL validation already defined somewhere.
+	// TODO: Validation to disallow unrealistic schedules (e.g */2 * * * * every 2 mins)
+	// +kubebuilder:validation:Optional
+	// +optional
+	Schedule string `json:"schedule"`
+
+	// The time zone name for the given schedule, see https://en.wikipedia.org/wiki/List_of_tz_database_time_zones.
+	// If not specified, this will default to the time zone of the kube-controller-manager process.
+	// See https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/#time-zones
+	// +kubebuilder:validation:Optional
+	// +optional
+	// TODO: CEL validation for timezone format
+	TimeZone string `json:"timeZone,omitempty"`
+
+	// RetentionPolicy defines the retention policy for retaining and deleting existing backups.
+	// +kubebuilder:validation:Optional
+	// +optional
+	RetentionPolicy RetentionPolicy `json:"retentionPolicy"`
+
+	// PVCName specifies the name of the PersistentVolumeClaim which binds a PersistentVolume where the
+	// etcd backup files would be saved
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +required
+	PVCName string `json:"pvcName"`
+}
+
+// RetentionType is the enumeration of valid retention policy types
+// +enum
+// +kubebuilder:validation:Enum:="RetentionCount";"RetentionSizeGb"
+type RetentionType string
+
+const (
+	RetentionTypeCount RetentionType = "RetentionCount"
+	RetentionTypeSize  RetentionType = "RetentionSizeGb"
+)
+
+// RetentionPolicy defines the retention policy for retaining and deleting existing backups.
+// This struct is a discriminated union that allows users to select the type of retention policy from the supported types.
+// +union
+type RetentionPolicy struct {
+	// RetentionType sets the type of retention policy. The currently supported and valid values are "retentionCount"
+	// Currently, the only valid policies are retention by count (RetentionCount) and by size (RetentionSizeGb). More policies or types may be added in the future.
+	// +unionDiscriminator
+	// +required
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum:=RetentionCount;RetentionSizeGb
+	RetentionType RetentionType `json:"retentionType"`
```

Comment:
I think this might still work based off the example of the controlplanemachine set union which also uses a type definition "alias".

https://github.com/openshift/api/blob/a6fbaf36017d235ff58a51ddc3962088dabbf6c3/machine/v1/types_controlplanemachineset.go#L122-L125

https://github.com/openshift/api/blob/a6fbaf36017d235ff58a51ddc3962088dabbf6c3/machine/v1/types_controlplanemachineset.go#L109-L114

But if that's not the convention or is ambiguous, I can update to a string alias declaration.

Reply:
> `type RetentionType string`

Looks like a string alias to me? Was this updated since the original comment?

Reply:
No I haven't changed it. I think `type RetentionType string` is what @ingvagabund was originally referring to as being "type definition" instead of being a "alias declaration" but I wasn't aware of the distinction before this.
https://go.dev/ref/spec#Type_declarations

In any case, it seems like a string type works as well and seems to be in accordance with the api convention guide:

> The discriminant MUST be a string (or string alias) type

Reply:
Ok, I think it's fine as it is. The bit that matters here is how it looks to the OpenAPI schema. When validations are applied, everything is unstructured, it won't know what the underlying Go type is, so whether it's a type declaration or an alias, it looks the same to the OpenAPI schema.